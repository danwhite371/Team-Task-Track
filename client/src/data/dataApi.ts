import type { Task } from '../types';
import {
  createTask,
  fetchTasks,
  startTask,
  stopTask,
  fetchTaskTimes,
} from './graphql';

export default class DataApi {
  updateTaskData: (tasks: Task[]) => void;
  tasks: Task[] = [];
  constructor(updateTaskData: (tasks: Task[]) => void) {
    this.updateTaskData = updateTaskData;
    this.sendTasks();
  }
  async createNewTask(name: string) {
    await createTask(name);
    await this.sendTasks();
  }
  async sendTasks() {
    const tasks = await fetchTasks();
    this.tasks = tasks;
    this.updateTaskData(tasks);
  }
  async startTask(id: number) {
    await startTask(id);
    await this.sendTasks();
  }

  async stopTask(id: number) {
    await stopTask(id);
    await this.sendTasks();
  }

  /*
    When updating task times, we want to get the data, update the task object, then send
    but we don't have the full list anymore as we just sent it.
  */
  async fetchTaskTimes(taskId: number) {
    const task = this.tasks.find((task) => task.id == taskId);
    if (!task) throw new Error('[sendTaskTimes] Task id mismatch');
    const taskTimes = await fetchTaskTimes(taskId);
    task.taskTimes = taskTimes;
    this.tasks = [...this.tasks];
    this.updateTaskData(this.tasks);
  }
}

/*
  We need a return object with status, error, message, data
  on init we begin sending tasks data to the UI and send a loading state to the UI

  Loading (status) a temp object
  data
  success message | error message - closable messages, and a message history

  On the UI we want to see the most recent messages and be able to see a message history
  So somewhere we have a list of message, that all have a read state

  We can have a separate message api to handle the messages, which is data coming in from the top level,
  and displayed somewhere in the UI. Let's start by passing everything down, then change is to a context.
*/
