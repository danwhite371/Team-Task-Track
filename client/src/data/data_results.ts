import type { OperationResult } from '@/types';

const LoadingLoading: OperationResult = {
  status: 'loading',
  message: 'Loading...',
};
const taskLoadingSuccess: OperationResult = {
  status: 'success',
  message: 'Tasks loaded.',
};

const loadingConnection: OperationResult = {
  status: 'loading',
  message: 'Connecting...',
};
const taskCreatedSuccess: OperationResult = {
  status: 'success',
  message: 'Task created.',
};
const taskCreatedError: OperationResult = {
  status: 'error',
  message: 'Task name cannot be empty.',
};

export {
  LoadingLoading,
  taskLoadingSuccess,
  loadingConnection,
  taskCreatedSuccess,
  taskCreatedError,
};
