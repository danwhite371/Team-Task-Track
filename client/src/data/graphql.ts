import { lowercaseFirstChar } from '@/until';
import {
  createTaskQuery,
  getAllTasksQuery,
  startTaskQuery,
  stopTaskQuery,
  getTaskTimesQuery,
} from './queries';
const GRAPHQL_URL = 'http://localhost:4000/';

type graphqlFetchProps = {
  query: string;
  operationName: string;
  variables?: any;
};
async function graphqlFetch(props: graphqlFetchProps) {
  console.log(`graphqlFetch\n${JSON.stringify(props, null, 2)}`);
  let responseData;
  try {
    const response = await fetch(GRAPHQL_URL, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify(props),
    });
    console.log(`response\n${JSON.stringify(response, null, 2)}`);
    responseData = await response.json();
    if (responseData.errors) {
      const message = responseData.errors[0].message;
      throw new Error(message);
    }
  } catch (error) {
    throw error;
  }

  console.log(`responseData\n${JSON.stringify(responseData, null, 2)}`);
  return responseData.data[lowercaseFirstChar(props.operationName)];
}

async function fetchTasks() {
  return await graphqlFetch({
    query: getAllTasksQuery,
    operationName: 'GetAllTasks',
  });
}

async function fetchTaskTimes(taskId: number) {
  const variables = { taskId };
  return await graphqlFetch({
    operationName: 'GetTaskTimes',
    query: getTaskTimesQuery,
    variables,
  });
}

async function createTask(name: string) {
  const variables = { name };
  return await graphqlFetch({
    operationName: 'CreateTask',
    query: createTaskQuery,
    variables,
  });
}

async function startTask(id: number) {
  const variables = { id };
  return await graphqlFetch({
    operationName: 'StartTask',
    query: startTaskQuery,
    variables,
  });
}

async function stopTask(id: number) {
  const variables = { id };
  return await graphqlFetch({
    operationName: 'StopTask',
    query: stopTaskQuery,
    variables,
  });
}

export { fetchTasks, fetchTaskTimes, createTask, startTask, stopTask };
