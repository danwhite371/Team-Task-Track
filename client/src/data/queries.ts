const taskQL = `id
      name
      active
      lastTime
      duration {
        milliseconds
        seconds
        minutes
        hours
        days
        years
      }
      secondsDuration
      createdAt
      updatedAt`;

const createTaskQuery = `
  mutation CreateTask($name: String!) {
    createTask(name: $name) {
      ${taskQL}
    }
  }
`;

const getTaskQuery = `
  query GetTask($id: Int!) {
    getTask(id: $id) {
      ${taskQL}
    }
  }
`;

const getAllTasksQuery = `query GetAllTasks {
  getAllTasks {
    ${taskQL}
  }
}`;

const startTaskQuery = `
  mutation StartTask($id: Int!) {
    startTask(id: $id) {
      ${taskQL}
    }
  }`;

const stopTaskQuery = `
  mutation StopTask($id: Int!) {
    stopTask(id: $id) {
      ${taskQL}
    }
  }`;

const deleteTaskQuery = `
  mutation DeleteTask($id: Int!) {
    deleteTask(id: $id) 
  }`;

const getTaskTimesQuery = `query GetTaskTimes($taskId: Int!){
    getTaskTimes(taskId: $taskId) {
      id
      start
      stop
      secondsDuration
    }
  }`;

const changeTaskNameQuery = `mutation ChangeTaskName($id: Int!, $name: String!) {
  changeTaskName(id: $id, name: $name) {
   name
    id
    id
    name
    active
    lastTime
    duration {
      milliseconds
      seconds
      minutes
      hours
      days
      years
    }
    secondsDuration
    createdAt
    updatedAt
  }
}`;

const queries = {
  createTaskQuery,
  getTaskQuery,
  getAllTasksQuery,
  startTaskQuery,
  stopTaskQuery,
  getTaskTimesQuery,
  deleteTaskQuery,
  changeTaskNameQuery,
};

export default queries;
