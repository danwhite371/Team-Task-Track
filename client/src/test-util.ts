function logMockCalls(mocked: jest.Mock<any, any, any>, label: string) {
  console.log(label);
  mocked.mock.calls.forEach((call, index) => {
    console.log(`Call ${index + 1}:`, call);
  });
}

function expectNthCalledWith(mocked: jest.Mock<any, any, any>, nth: number, containing: any) {
  expect(mocked).toHaveBeenNthCalledWith(nth, expect.objectContaining(containing));
}

export { logMockCalls, expectNthCalledWith };
