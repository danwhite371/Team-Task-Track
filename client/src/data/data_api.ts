import type { OperationResult, Task } from '../types';
import { loadingConnection, LoadingLoading } from './data_results';
import {
  createTask,
  fetchTasks,
  startTask as startTaskRequest,
  stopTask as stopTaskRequest,
  fetchTaskTimes as fetchTaskTimesRequest,
} from './graphql';

export default class DataApi {
  updateTaskData: (tasks: Task[]) => void;
  updateOperationResult: (operationResult: OperationResult) => void;
  tasks: Task[] = [];

  constructor(
    updateTaskData: (tasks: Task[]) => void,
    updateOperationResult: (operationResult: OperationResult) => void
  ) {
    this.updateTaskData = updateTaskData;
    this.updateOperationResult = updateOperationResult;
    this.sendTasks();
  }

  async createNewTask(name: string) {
    this.updateOperationResult(loadingConnection);
    try {
      const task = await createTask(name);
      console.log('createNewTask', JSON.stringify(task, null, 2));
      this.tasks.unshift(task);
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task created.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  async sendTasks() {
    this.updateOperationResult(LoadingLoading);
    try {
      const tasks = await fetchTasks();
      this.tasks = tasks;
      this.updateTaskData(tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Tasks loaded.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  // Get a new Task item with, replace current item with it, sort items by last time
  async startTask(id: number) {
    console.log('[DataApi] startTask', id);
    this.updateOperationResult(loadingConnection);
    try {
      const task = await startTaskRequest(id);
      this.replaceTask(id, task);
      this.sortTasks();
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task timer starter.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  async stopTask(id: number) {
    console.log('[DataApi] stopTask', id);
    this.updateOperationResult({ status: 'loading', message: 'Connecting...' });
    try {
      const task = await stopTaskRequest(id);
      console.log('[DataApi] stopTask', JSON.stringify(task, null, 2));
      this.replaceTask(id, task);
      this.sortTasks();
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task timer stopped.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  replaceTask(id: number, task: Task) {
    const i = this.tasks.findIndex((task) => task.id === id);
    console.log('[DataApi] replaceTask', this.tasks[i], task);
    this.tasks[i] = task;
  }

  sortTasks() {
    this.tasks.sort(
      (taskA, taskB) => Number(taskB.lastTime) - Number(taskA.lastTime)
    );
  }

  async fetchTaskTimes(taskId: number) {
    console.log('[DataApi] fetchTaskTimes');
    this.updateOperationResult({ status: 'loading', message: 'Loading...' });
    try {
      const task = this.tasks.find((task) => task.id == taskId);
      if (!task) throw new Error('[sendTaskTimes] Task id mismatch');
      const taskTimes = await fetchTaskTimesRequest(taskId);
      task.taskTimes = taskTimes;
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task times loaded.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }
}
