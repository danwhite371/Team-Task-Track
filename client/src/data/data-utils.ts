import type { DataUtils } from '@/types';
import queries from './queries';

const { createTaskQuery, getAllTasksQuery, getTaskTimesQuery, startTaskQuery, stopTaskQuery } = queries;

export const dataUtils: DataUtils = {
  results: {
    loadingLoading: {
      status: 'loading',
      message: 'Loading...',
    },
    taskLoadingSuccess: {
      status: 'success',
      message: 'Tasks loaded.',
    },
    taskTimesLoadingSuccess: {
      status: 'success',
      message: 'Task times loaded.',
    },
    loadingConnection: {
      status: 'loading',
      message: 'Connecting...',
    },
    taskCreatedSuccess: {
      status: 'success',
      message: 'Task created.',
    },
    startTaskSuccess: {
      status: 'success',
      message: 'Task timer starter.',
    },
    stopTaskSuccess: {
      status: 'success',
      message: 'Task timer stopped.',
    },
    taskCreatedError: {
      status: 'error',
      message: 'Task name cannot be empty.',
    },
    failedFetchError: {
      status: 'error',
      message: 'Failed to fetch',
    },
  },
  requests: {
    getAllTasks: {
      operationName: 'GetAllTasks',
      query: getAllTasksQuery,
    },
    createTask: (name) => ({
      operationName: 'CreateTask',
      query: createTaskQuery,
      variables: { name },
    }),
    getTaskTimes: (taskId) => ({
      operationName: 'GetTaskTimes',
      query: getTaskTimesQuery,
      variables: { taskId },
    }),
    startTask: (id) => ({
      operationName: 'StartTask',
      query: startTaskQuery,
      variables: { id },
    }),
    stopTask: (id) => ({
      operationName: 'StopTask',
      query: stopTaskQuery,
      variables: { id },
    }),
  },
};
