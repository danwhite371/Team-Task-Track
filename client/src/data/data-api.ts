import type { OperationResult, Task } from '../types';
import { dataUtils } from './data-utils';
import {
  createTask,
  fetchTasks,
  startTask as startTaskRequest,
  stopTask as stopTaskRequest,
  fetchTaskTimes as fetchTaskTimesRequest,
} from './graphql';
const { results } = dataUtils;

interface UpdateTaskData {
  (tasks: Task[]): void;
}

interface UpdateOperationResult {
  (operationResult: OperationResult): void;
}

export default class DataApi {
  updateTaskData: UpdateTaskData;
  updateOperationResult: UpdateOperationResult;
  tasks: Task[] = [];

  private constructor(
    updateTaskData: UpdateTaskData,
    updateOperationResult: UpdateOperationResult
  ) {
    this.updateTaskData = updateTaskData;
    this.updateOperationResult = updateOperationResult;
  }

  static async create(
    updateTaskData: UpdateTaskData,
    updateOperationResult: UpdateOperationResult
  ): Promise<DataApi> {
    const dataApi = new DataApi(updateTaskData, updateOperationResult);
    await dataApi.sendTasks();
    return dataApi;
  }

  async createNewTask(name: string) {
    this.updateOperationResult(results.loadingConnection);
    try {
      const task = await createTask(name);
      console.log('createNewTask', JSON.stringify(task, null, 2));
      this.tasks.unshift(task);
      this.tasks = [...this.tasks];
      this.updateTaskData([...this.tasks]);
      this.updateOperationResult(results.taskCreatedSuccess);
    } catch (error: unknown) {
      this.updateOperationResult(results.taskCreatedError);
    }
  }

  async sendTasks() {
    this.updateOperationResult(results.loadingLoading);
    try {
      const tasks = await fetchTasks();
      this.tasks = tasks;
      this.updateTaskData([...tasks]);
      this.updateOperationResult(results.taskLoadingSuccess);
    } catch (error: any) {
      this.updateOperationResult({
        status: 'error',
        message: error.message,
      });
    }
  }

  // Get a new Task item with, replace current item with it, sort items by last time
  async startTask(id: number) {
    console.log('[DataApi] startTask', id);
    this.updateOperationResult(results.loadingConnection);
    try {
      const task = await startTaskRequest(id);
      this.replaceTask(id, task);
      this.sortTasks();
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult(results.startTaskSuccess);
      console.log('after startTask updateOperationResult');
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  async stopTask(id: number) {
    console.log('[DataApi] stopTask', id);
    this.updateOperationResult(results.loadingConnection);
    try {
      const task = await stopTaskRequest(id);
      console.log('[DataApi] stopTask', JSON.stringify(task, null, 2));
      this.replaceTask(id, task);
      this.sortTasks();
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult(results.stopTaskSuccess);
      console.log('after stopTask updateOperationResult');
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  replaceTask(id: number, task: Task) {
    const i = this.tasks.findIndex((task) => task.id === id);
    console.log('[DataApi] replaceTask', this.tasks[i], task);
    this.tasks[i] = task;
  }

  sortTasks() {
    this.tasks.sort(
      (taskA, taskB) => Number(taskB.lastTime) - Number(taskA.lastTime)
    );
  }

  async fetchTaskTimes(taskId: number) {
    console.log('[DataApi] fetchTaskTimes');
    this.updateOperationResult(results.loadingLoading);
    try {
      const task = this.tasks.find((task) => task.id == taskId);
      if (!task) throw new Error('[sendTaskTimes] Task id mismatch');
      const taskTimes = await fetchTaskTimesRequest(taskId);
      task.taskTimes = taskTimes;
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult(results.taskTimesLoadingSuccess);
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }
}
