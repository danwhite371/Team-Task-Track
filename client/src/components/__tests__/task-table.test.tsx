import { act, cleanup, fireEvent, render, screen, waitFor, within } from '@testing-library/react';
import TaskTable from '../task-table';
import { tasks, allTaskTimes } from '../../data/__tests__/data-test-util';
import DataApi from '@/data/data-api';
import mockFetch from '../../data/__mocks__/fetch';
import { dataUtils } from '@/data/data-utils';
import { expectNthCalledWith } from '@/test-util';

const { results } = dataUtils;
global.fetch = mockFetch;

const mockUpdateTaskData = jest.fn();
const mockUpdateOperationResult = jest.fn();

beforeEach(() => {
  jest.useFakeTimers();
  jest.clearAllMocks();
});

afterEach(() => {
  jest.useRealTimers();
  cleanup();
});

describe('TaskTable', () => {
  it('should render a task table', async () => {
    const DURATION = 1;
    const NAME = 0;
    const START_TIME = 0;
    const dataApi = await DataApi.create(mockUpdateTaskData, mockUpdateOperationResult);
    expect(tasks).toMatchSnapshot();
    const { rerender } = render(<TaskTable tasks={tasks} dataApi={dataApi} />);
    await waitFor(() => {
      expectNthCalledWith(mockUpdateOperationResult, 2, results.taskLoadingSuccess);
    });
    jest.clearAllMocks();

    for (const task of tasks) {
      const row = screen.getByTestId(`${task.id}`);
      const cells = within(row).getAllByRole('cell');
      expect(cells[NAME]).toHaveTextContent(task.name);
      if (task.active) {
        // check that the stop button is clicked and that fetch is called
        const button = within(row).queryByRole('button', { name: /stop/i });
        expect(button).not.toBeNull();
        act(() => {
          fireEvent.click(button!);
        });
        await waitFor(() => {
          expectNthCalledWith(mockUpdateOperationResult, 2, results.stopTaskSuccess);
        });
        jest.clearAllMocks();
      } else {
        // check that the start button is clicked and that fetch is called
        const button = within(row).queryByRole('button', { name: /start/i });
        expect(button).not.toBeNull();
        act(() => {
          fireEvent.click(button!);
        });
        await waitFor(() => {
          expectNthCalledWith(mockUpdateOperationResult, 2, results.startTaskSuccess);
        });
        jest.clearAllMocks();
      }
      // if we have at least one task time row for this task
      if (task.active || task.duration) {
        expect(cells[DURATION]).not.toBeEmptyDOMElement();
        // if we don't have task times for this task, throw an error
        // then click on task name to load time times
        const taskTimes = allTaskTimes.filter((tt) => tt.taskId == task.id);
        if (!taskTimes) throw new Error(`expected taskTimes not found for taskId ${task.id}`);
        task.taskTimes = taskTimes;
        act(() => {
          fireEvent.click(cells[NAME]);
        });

        await waitFor(() => {
          expectNthCalledWith(mockUpdateOperationResult, 2, results.taskTimesLoadingSuccess);
        });
        // logMockCalls(mockUpdateOperationResult, 'mockUpdateOperationResult: line 63');
        rerender(<TaskTable tasks={tasks} dataApi={dataApi} />);
        // check that we have a start time for each task time row
        const timeTestId = `time-${task.id}-${taskTimes.length - 1}`;
        console.log('timeTestId', timeTestId);
        const lastRow = await screen.findByTestId(timeTestId);
        const rowCells = within(lastRow).getAllByRole('cell');
        expect(rowCells[START_TIME]).not.toBeEmptyDOMElement();
        jest.clearAllMocks();
      } else {
        // This task should have no duration
        expect(cells[DURATION]).toBeEmptyDOMElement();
      }
    }
  }, 10000);
});
