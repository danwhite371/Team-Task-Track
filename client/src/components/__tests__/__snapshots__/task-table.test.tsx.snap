// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTable should render a task table 1`] = `
[
  {
    "active": true,
    "createdAt": "1750551075680",
    "duration": {
      "days": null,
      "hours": null,
      "milliseconds": 835,
      "minutes": null,
      "seconds": 4,
      "years": null,
    },
    "id": 95,
    "lastTime": "1750551083234",
    "name": "Active Duration with times test",
    "secondsDuration": 4.835,
    "taskTimes": [],
    "updatedAt": "1750551075680",
  },
  {
    "active": true,
    "createdAt": "1750551035076",
    "duration": null,
    "id": 94,
    "lastTime": "1750551040643",
    "name": "Active Duration test",
    "secondsDuration": null,
    "taskTimes": [],
    "updatedAt": "1750551035076",
  },
  {
    "active": false,
    "createdAt": "1750168859679",
    "duration": {
      "days": null,
      "hours": null,
      "milliseconds": 779,
      "minutes": 55,
      "seconds": 20,
      "years": null,
    },
    "id": 85,
    "lastTime": "1750551009498",
    "name": "TimeTable component test",
    "secondsDuration": 3320.779,
    "taskTimes": [],
    "updatedAt": "1750168859679",
  },
  {
    "active": false,
    "createdAt": "1750546046133",
    "duration": null,
    "id": 1,
    "lastTime": "1750546046142",
    "name": "Test 2",
    "secondsDuration": null,
    "taskTimes": [],
    "updatedAt": "1750546046142",
  },
  {
    "active": false,
    "createdAt": "1750541047569",
    "duration": null,
    "id": 93,
    "lastTime": "1750541047569",
    "name": "No duration test",
    "secondsDuration": null,
    "taskTimes": [],
    "updatedAt": "1750541047569",
  },
  {
    "active": true,
    "createdAt": "1750339142100",
    "duration": {
      "days": 1,
      "hours": 2,
      "milliseconds": 593,
      "minutes": 26,
      "seconds": 16,
      "years": null,
    },
    "id": 92,
    "lastTime": "1750434330601",
    "name": "Long test",
    "secondsDuration": 95176.593,
    "taskTimes": [],
    "updatedAt": "1750339142100",
  },
  {
    "active": false,
    "createdAt": "1750338486924",
    "duration": {
      "days": 1,
      "hours": 2,
      "milliseconds": 451,
      "minutes": 34,
      "seconds": 44,
      "years": null,
    },
    "id": 91,
    "lastTime": "1750434187660",
    "name": "Test 2",
    "secondsDuration": 95684.451,
    "taskTimes": [],
    "updatedAt": "1750338486924",
  },
  {
    "active": false,
    "createdAt": "1750338481645",
    "duration": {
      "days": null,
      "hours": null,
      "milliseconds": 661,
      "minutes": 1,
      "seconds": 43,
      "years": null,
    },
    "id": 90,
    "lastTime": "1750338604570",
    "name": "Test 1",
    "secondsDuration": 103.661,
    "taskTimes": [],
    "updatedAt": "1750338481645",
  },
  {
    "active": false,
    "createdAt": "1750337122976",
    "duration": {
      "days": null,
      "hours": null,
      "milliseconds": 69,
      "minutes": 18,
      "seconds": 1,
      "years": null,
    },
    "id": 89,
    "lastTime": "1750338209903",
    "name": "GraphQL - Return all tasks with task times",
    "secondsDuration": 1081.069,
    "taskTimes": [],
    "updatedAt": "1750337122976",
  },
  {
    "active": false,
    "createdAt": "1750168888975",
    "duration": {
      "days": null,
      "hours": 1,
      "milliseconds": 1,
      "minutes": 47,
      "seconds": 12,
      "years": null,
    },
    "id": 86,
    "lastTime": "1750295937277",
    "name": "ActiveDuration component test",
    "secondsDuration": 6432.001,
    "taskTimes": [],
    "updatedAt": "1750168888975",
  },
  {
    "active": false,
    "createdAt": "1747611904169",
    "duration": {
      "days": null,
      "hours": 1,
      "milliseconds": 747,
      "minutes": 17,
      "seconds": 40,
      "years": null,
    },
    "id": 39,
    "lastTime": "1747697369880",
    "name": "Integrate logger with Sequelize",
    "secondsDuration": 4660.747,
    "taskTimes": [],
    "updatedAt": "1747611904169",
  },
  {
    "active": false,
    "createdAt": "1747531067029",
    "duration": {
      "days": null,
      "hours": null,
      "milliseconds": 205,
      "minutes": 42,
      "seconds": 47,
      "years": null,
    },
    "id": 31,
    "lastTime": "1747533641016",
    "name": "Add logging",
    "secondsDuration": 2567.205,
    "taskTimes": [],
    "updatedAt": "1747531067029",
  },
  {
    "active": false,
    "createdAt": "1747483574242",
    "duration": {
      "days": null,
      "hours": null,
      "milliseconds": 7,
      "minutes": 25,
      "seconds": 14,
      "years": null,
    },
    "id": 29,
    "lastTime": "1747485093763",
    "name": "Clean up any personal info",
    "secondsDuration": 1514.007,
    "taskTimes": [],
    "updatedAt": "1747483574242",
  },
  {
    "active": false,
    "createdAt": "1747348099047",
    "duration": {
      "days": null,
      "hours": 2,
      "milliseconds": 326,
      "minutes": 34,
      "seconds": 18,
      "years": null,
    },
    "id": 28,
    "lastTime": "1747445673153",
    "name": "Convert backend to TypeScript",
    "secondsDuration": 9258.326,
    "taskTimes": [],
    "updatedAt": "1747348099047",
  },
  {
    "active": false,
    "createdAt": "1746883377114",
    "duration": {
      "days": null,
      "hours": 1,
      "milliseconds": 878,
      "minutes": 53,
      "seconds": 27,
      "years": null,
    },
    "id": 10,
    "lastTime": "1747324842941",
    "name": "Planning",
    "secondsDuration": 6807.878,
    "taskTimes": [],
    "updatedAt": "1746883377114",
  },
  {
    "active": false,
    "createdAt": "1747086364931",
    "duration": {
      "days": null,
      "hours": 2,
      "milliseconds": 597,
      "minutes": 15,
      "seconds": 46,
      "years": null,
    },
    "id": 21,
    "lastTime": "1747155552005",
    "name": "feat(time): Create component to render a time interval",
    "secondsDuration": 8146.597,
    "taskTimes": [],
    "updatedAt": "1747086364931",
  },
]
`;
